waveorder/waveorder_reconstructor.py

```diff
@@ -977,23 +988,40 @@ def Stokes_recon(self, I_meas):
         Parameters
         ----------
             I_meas        : numpy.ndarray
-                            polarization-sensitive intensity images with the size of (N_channel, ...)
+                            polarization-sensitive intensity images with the size of (N_channel, N, M) or
+                            (N_channel, N, M, N_defocus)
                           
         Returns
         -------
             S_image_recon : numpy.ndarray
-                            reconstructed Stokes parameters with the size of (N_Stokes, ...)
+                            reconstructed Stokes parameters with the size of (N_Stokes, N, M) or
+                            (N_channel, N, M, N_defocus)
                        
                               
         '''
-        
-        img_shape = I_meas.shape
-        
-        A_pinv = np.linalg.pinv(self.A_matrix)
-        S_image_recon = np.reshape(np.dot(A_pinv, I_meas.reshape((self.N_channel, -1))), (self.N_Stokes,)+img_shape[1:])
 
-            
-        return S_image_recon
+        n_dims = I_meas.ndim
+        if n_dims not in (3, 4):
+            raise ValueError('Unsupported image data size. Image data must be of size (N_channel, N, M) or '
+                             '(N_channel, N, M, N_defocus)')
+
+        if n_dims == 3:
+            I_meas = I_meas[..., np.newaxis]
+        img_data = np.moveaxis(I_meas, 0, -2)
+
+        if self.use_gpu:
+            if self._A_matrix_inv_gpu_array is None:
+                self._A_matrix_inv_gpu_array = cp.array(self.A_matrix_inv)
```

Comment:
Question -- do we need to keep the `_A_matrix_inv_gpu_array` as a class attribute?  It doesn't look like it is ever called again outside of this function.  Perhaps we can leave it as a temporary variable within this function?

Reply:
I reuse the `_A_matrix_inv_gpu_array` in repeated calls to `Stokes_recon`, but it's true that it is not used by other methods. Do you have another way to handle that?

Reply:
Ok yeah this makes sense then, I think this is a fine way of handling it.